File: C++/lowestCommonAncestor.cpp
#include<iostream>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    int lowestCommonAncestor(TreeNode* root, int p, int q) {
        while(root){
            if(p<root->val&&q<root->val){
                root=root->left;
            }
            else if(p>root->val&&q>root->val){
                root =root->right;
            }
            return root->val;
        }
        return 0;
    }
};

void deleteTree(TreeNode* root){
    if(root){
        deleteTree(root->left);
        deleteTree(root->right);
        delete root;
    }
}

int main(){
    TreeNode* root=new TreeNode(7);
    root->left=new TreeNode(1);
    root->right=new TreeNode(12);
    root->left->left=new TreeNode(0);
    root->left->right=new TreeNode(4);
    root->right->left=new TreeNode(11);
    root->right->right=new TreeNode(14);
    root->left->right->left=new TreeNode(3);
    root->left->right->right=new TreeNode(5);
    Solution solution;
    int p=1;
    int q=12;

    int commonancestor=solution.lowestCommonAncestor(root,p,q);
    std::cout<<commonancestor<<std::endl;
    deleteTree(root);
    return 0;
}

File: python/lowestCommonAncestor.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self , root: TreeNode, p: int, q: int) -> int:
        while root:
            if p<root.val and q<root.val:
                root=root.left
            if p>root.val and q>root.val:
                root=root.right
            return root.val
        return 0
    
root=TreeNode(7)
root.left=TreeNode(1)
root.right=TreeNode(12)
root.left.left=TreeNode(0)
root.left.right=TreeNode(4)
root.right.left=TreeNode(11)
root.right.right=TreeNode(14)
root.left.right.left=TreeNode(3)
root.left.right.right=TreeNode(5)
p=3
q=11
solution=Solution()
common_ancestor=solution.lowestCommonAncestor(root,p,q)
print(common_ancestor)